docs/proposals/initializers.md

```diff
@@ -0,0 +1,110 @@
+
+*This document introduces the concept of initializers in Kubernetes, and describes use cases and implementation details.*
+
+## Motivation
+
+The ability to introduce hooks into the API to modify an object prior to it being acted upon in the system is a common requirement.
+
+The primary mechanism to achieve this use-case today is to write an "admission controller" which is a server-side filter that observes API server actions, and provides the ability for an ordered set of filters to observe, modify, accept, or reject an object from being persisted into the API server.  By providing the ability to give an immediate yes/no response to the user, it provides a good place for hooks for use-cases that require an immediate user response.  For example, the user is able to be told that they are not able to create an object in a namespace without first creating that namespace.  Another common example is quota enforcement where the user is immediately forbidden from exceeding any quota constraints.  In essence, its a synchronous visitor pattern.
+
+While powerful, admission controllers have drawbacks.  The major drawback is that it requires a recompilation and restart of the API server to introduce new behavior.  For those scenarios where an immediate yes/no response was not required for the user experience, it would be ideal if the system could support an asynchronous visitor pattern that would allow controllers external to the API server to observe, and modify a resource prior to it becoming 'active' in the system.  This would allow a new type of controller to be authored in the system that can initialize a resource similar to how other controllers watch and act upon already initialized objects.  There are many scenarios where this would be appropriate.  For example, a pod auto-sizer could be deployed separate from the API server that watches for incoming pods.  The pod-autosizer should be able to assign compute resource requests prior to the pod being scheduled.  If authored as an "initializer", it would allow anyone to bring their own logic to the cluster for how sizing occurs without needing to recompile or change the core components.
+
+## Goals
+
+1. Enable asynchronous controller pattern (i.e. initializers) as a viable alternative to existing admission control to support loose coupling of components where synchronous user response is less critical.
+2. Enable registration of a set of initializers to an API resource that must be able to act on an object to support its initialization.
+
+## Model changes
+
+**Initializer**
+
+A new type introduced into the API that allows the registration of an initializer with an associated group and kind in the API.  This allows dynamic registration of initializers without a forced restart or recompilation of the server.
+
+**NOTE: It's possible we could support this without introduction of a new type, maybe via the ConfigData resource, needs thought and input from community**
+
+**ObjectMeta.Initializers**
```

Comment:
It might make sense to more loosely structure dependencies - they are
closer to annotations to me, and I'd like to be able to attach limited
metadata with some dependencies (for instance, representing triggers that
need an object reference and a few bits of data).  We can serialize it as
an annotation, but then generic clients lose the directed edge relationship.


On Nov 16, 2015, at 7:42 PM, Brian Grant <notifications@github.com> wrote:

In docs/proposals/initializers.md
<https://github.com/kubernetes/kubernetes/pull/17305#discussion_r45007826>:

> +While powerful, admission controllers have drawbacks.  The major drawback is that it requires a recompilation and restart of the API server to introduce new behavior.  For those scenarios where an immediate yes/no response was not required for the user experience, it would be ideal if the system could support an asynchronous visitor pattern that would allow controllers external to the API server to observe, and modify a resource prior to it becoming 'active' in the system.  This would allow a new type of controller to be authored in the system that can initialize a resource similar to how other controllers watch and act upon already initialized objects.  There are many scenarios where this would be appropriate.  For example, a pod auto-sizer could be deployed separate from the API server that watches for incoming pods.  The pod-autosizer should be able to assign compute resource requests prior to the pod being scheduled.  If authored as !
 an "
 initializer", it would allow anyone to bring their own logic to the
cluster for how sizing occurs without needing to recompile or change
the core components.
> +
> +## Goals
> +
> +1. Enable asynchronous controller pattern (i.e. initializers) as a viable alternative to existing admission control to support loose coupling of components where synchronous user response is less critical.
> +2. Enable registration of a set of initializers to an API resource that must be able to act on an object to support its initialization.
> +
> +## Model changes
> +
> +**Initializer**
> +
> +A new type introduced into the API that allows the registration of an initializer with an associated group and kind in the API.  This allows dynamic registration of initializers without a forced restart or recompilation of the server.
> +
> +**NOTE: It's possible we could support this without introduction of a new type, maybe via the ConfigData resource, needs thought and input from community**
> +
> +**ObjectMeta.Initializers**

Concretely, using Pod as an example:

type Pod struct {
    unversioned.TypeMeta `json:",inline"`
    ObjectMeta `json:"metadata,omitempty"`
    Dependencies ObjectDependencies `json:"dependencies,omitempty"`
    Spec PodSpec `json:"spec,omitempty"`
    Status PodStatus `json:"status,omitempty"`
}
type ObjectDependencies struct {
    Initializers map[string]string `json:"initializers,omitempty"`
    Finalizers map[string]string `json:"finalizers,omitempty"`
    ExistenceDependencies []ObjectReference
`json:"existenceDependencies,omitempty"`
    ControllerRef *ObjectReference `json:"controllerRef,omitempty"`
...
}

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17305/files#r45007826>.